### ISSUE TITLE
support for eslint `no-restricted-syntax`

### ISSUE THREAD

**Issue by dnim** [oxc-project/oxc#7342]
Hi, 

we actively using custom selectors with [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax) eslint rule. 

Are the any plans to support this feature in the nearest future? 

Thanks!

**Comment by Boshen** [comment]
We are unable to support this rule. Try https://github.com/ast-grep/ast-grep instead.

Which syntax are you trying to restrict? We have a few implementations such as https://oxc.rs/docs/guide/usage/linter/rules/oxc/no-rest-spread-properties.html or https://oxc.rs/docs/guide/usage/linter/rules/oxc/no-optional-chaining.html

**Comment by dnim** [comment]
Hi @Boshen! 

Thanks you for the fast reply 🙇 

currently we using underscore detection rule for the import, i.e. 
`"selector": "ImportSpecifier[imported.name=/^_/]"` 

`ast-grep` seems a really good fit for that task, thanks.

**Comment by klippx** [comment]
We would like to ban enums:

```
    "no-restricted-syntax": [
      "error",
      {
        "selector": "TSEnumDeclaration",
        "message": "You should not use TypeScript enums, according to the Engineering Handbook."
      }
    ],
```

**Comment by camc314** [comment]
> We would like to ban enums:
> 
> ```
>     "no-restricted-syntax": [
>       "error",
>       {
>         "selector": "TSEnumDeclaration",
>         "message": "You should not use TypeScript enums, according to the Engineering Handbook."
>       }
>     ],
> ```

This could be implemented as a custom plugin:

```
import type { Plugin } from 'oxlint';

const plugin: Plugin = {
  meta: {
    name: 'basic-custom-plugin',
  },
  rules: {
    'no-debugger': {
      create(context) {
        return {
          TSEnumDeclaration(enumDeclaration) {
            context.report({
              message: 'Avoid using enum declarations.\nYou should not use TypeScript enums, according to the Engineering Handbook.',
              node: enumDeclaration,
            });
          },
        };
      },
    },
  },
};

export default plugin;

```

**Comment by overlookmotel** [comment]
We may be able to support this rule in its entirety by running it as a JS rule. I'm going to look into that in the next week.

**Comment by angrybacon** [comment]
For posterity and if anyone is looking for more usage example, here is how we disallow `Props` exports

```ts
import { type Rule } from '@oxlint/plugins';

export const noUbiquitousExport: Rule = {
  createOnce: (context) => ({
    ExportNamedDeclaration(node) {
      if (
        (node.declaration?.type === 'TSInterfaceDeclaration' ||
          node.declaration?.type === 'TSTypeAliasDeclaration') &&
        node.declaration.id.name === 'Props'
      ) {
        context.report({
          message:
            'Exporting "Props" is not allowed. Prefer a less generic name.',
          node: node.declaration.id,
        });
      }
    },
  }),
};
```

**Comment by camc314** [comment]
For anyone finding this issue in the future, please look at https://oxc.rs/blog/2026-03-11-oxlint-js-plugins-alpha.html#eslint-rules

1. add `oxlint-plugin-eslint` dependency
2. configure your `.oxlintrc.json` like so:

```json
{
  "jsPlugins": ["oxlint-plugin-eslint"],
  "rules": {
    // Note: "eslint-js" not "eslint"
    "eslint-js/no-restricted-syntax": [
      "error",
      {
        "selector": "ThrowStatement > CallExpression[callee.name=/Error$/]",
        "message": "Use `new` keyword when throwing an `Error`.",
      },
    ],
  },
}
```

3. run oxlint


This uses eslint's version of the implementation of the rule, allowing for full compat with `no-restricted-syntax`

Please reach out if you have any issues!

**Comment by angrybacon** [comment]
@camc314 I'm very much not familiar with the internals of either approach, is `oxlint-plugin-eslint` provided for convenience in a progressive migration or is it as efficient as the custom examples provided in this issue? Assuming of course they all use `createOnce`

**Comment by camc314** [comment]
> @camc314 I'm very much not familiar with the internals of either approach, is oxlint-plugin-eslint provided for convenience in a progressive migration or is it as efficient as the custom examples provided in this issue? Assuming of course they all use createOnce

`oxlint-plugin-eslint` is provided as a way to run the eslint rules in oxlint that oxlint does not yet nativly support. It probably won't be as efficient/performant as your example, as it is just a direct-re-export of eslint's rule with no additional optimizations. 

Given this, the package is just provided as a convenience to assist with users migrating to oxlint to fill in the missing oxlint rules.

Thanks!
